gateway/handler_unixfs_dir.go

```diff
@@ -59,20 +64,21 @@ func (i *handler) serveDirectory(ctx context.Context, w http.ResponseWriter, r *
 		}
 	}
 
+	// TODO: Was there a reason why this index.html check came after the redirect above that makes the inversion here incorrect?
```

Comment:
My understanding is this is both a security and UX/DX feature.

- UX: by normalizing to `/` first we avoid having duplicated entries for in browser history. Without this, `index.html` would render just fine on both `/ipfs/cid/foo` and `/ipfs/cid/foo/`, polluting web crawler indexes, user browser history etc with duplicate entries.
- DX/Security: There are things on the web platform that rely on path segments to define the scope of certain operations.  By doing redirect first, and ensuring directories end with `/`, we created deterministic behavior, so JS does not break in funny ways.
  - Oner example is [ServiceWorker which is registered for specific path-based "scope"](https://w3c.github.io/ServiceWorker/#document-context), so the same `index.html` could either be in some scopes, or not, depending on if URL had `/`
  - There is somehow related security nuance described in  [handler.go#L857-L864](https://github.com/ipfs/go-libipfs/blob/36918f45f2609be66219b0149d47053f3d955b97/gateway/handler.go#L857-L864)  – if we allowed for service worker on `/ipfs/cid`,  it could hijack requests to `/ipfs/*` and MITM gateway users. But we block that, and directories always end with `/` so its at best limited to specific `/ipfs/cid/*`

Reply:
Thanks, that's helpful. For now shouldn't have to change anything, but dealing with this will add some complexity if we allow externalizing index.html to the API implementation and letting implementations prefetch content like index.html pages.